sklearn/cluster/_hierarchical_fast.pyx

```diff
@@ -354,8 +366,7 @@ cdef class UnionFind(object):
         return n
 
 
-cpdef cnp.ndarray[DTYPE_t, ndim=2] _single_linkage_label(
-    cnp.ndarray[DTYPE_t, ndim=2] L):
+def _single_linkage_label(cnp.float64_t[:, :] L):
```

Comment:
What is the reason for using `def` and not `cpdef` with output type?

```suggestion
cpdef DTYPE_t[:, :] _single_linkage_label(const cnp.float64_t[:, :] L):
```

`L` seems also to be constant.

Reply:
I think this forces the output to be a memory view whereas we want to return the actual array.

Reply:
OK I see. @jjerphan What is the Cython API to then acknowledge that you want to return a NumPy array and not a memory view?

Reply:
One can use `cnp.ndarray`, but it is better not to depend on this interface if we do not need to.

For functions that are only called from Python code, there's little value of using return type because everything is seen as an object in Python. This is the case here.